MockedRethinkDB.js

```diff
@@ -0,0 +1,334 @@
+        }
+      });
+    });
+  }
+
+  /**
+   * Opens an active connection to the mocked database.
+   * @return {Promise<Connection>} resolves with a RethinkDB connection.
+  */
+  connection() {
+    if(this._connection) { return this._connection; }
+    return this._connection = this
+      .hackModule()
+      .then(() => require("rethinkdb"));
+  }
+
+  /**
+   * Creates the databases included in the initial data.
+   * @return {Promise} resolves when all databases have been created.
+   * @private
+  */
+  createDatabases() {
+    if(this._createDatabases) { return this._createDatabases; }
+    return this._createDatabases = Promise
+      .all([
+        this.hackModule(),
+        this.connection(),
+        Object.keys(this.initialDatabases()),
+      ])
+      .then(([ r, conn, dbs ]) => Promise.all( dbs.map(db => r.dbCreate(db).run(conn)) ));
+  }
+
+  /**
+   * Create the tables included in the initial data.
+   * @return {Promise} resolves when all tables have been created.
+   * @private
+  */
+  createTables() {
+    if(this._createTables) { return this._createTables; }
+    return this._createTables = Promise
+      .all([
+        this.hackModule(),
+        this.connection(),
+        this.initialTables(),
+        this.createDatabases(),
+      ])
+      .then(([ r, conn, tables ]) => Promise.all( tables.map(({ db, table }) => {
+        return r.db(db).tableCreate(table).run(conn);
+      })));
+  }
+
+  /**
+   * Create the documents included in the initial data.
+   * @return {Promise} resolves when all the documents have been created.
+   * @private
+  */
+  createDocuments() {
+    if(this._createDocs) { return this._createDocs; }
+    return this._createDocs = Promise
+      .all([
+        this.hackModule(),
+        this.connection(),
+        this.initialTables(),
+        this.createTables(),
+      ])
+      .then(([ r, conn, docs ]) => Promise.all( docs.map(({ db, table, doc }) => {
+        return r.db(db).table(table).insert(doc).run(conn);
+      })));
+  }
+
+  /**
+   * Filters the initial databases to exclude test databases.
+   * @return {Array<Object>} the valid databases.
+   * @private
+  */
+  initialDatabases() {
+    return filter(this.initialData, (val, index) => index !== "test");
+  }
+
+  /**
+   * Reduces the tables from all of the inital databases into `[ { db: "name", table: "name", docs: [ ... ] } ]`.
+   * @return {Array} each of the tables in the inital data
+   * @private
+  */
+  initialTables() {
+    function flattenDatabase(tables, db) {
+      return map(tables, (docs, table) => { db, table, docs });
+    }
+    return flatMap(this.initialDatabases(), flattenDatabase);
+  }
+
+  /**
+   * Reduces the documents from all tables in the initial databases into
+   * ` [ { db: "name", table: "name", doc: { ... } } ]`
+   * @return {Array} each of the tables in the initial data.
+   * @private
+  */
+  initialDocuments() {
+    function flattenTable({ db, table, docs }) {
```

Comment:
Can we define this outside of this method? Defining it inside will create this function every time this method is called.